delphi/umap_narrative/reset_conversation.py

```diff
@@ -38,6 +39,177 @@ def get_boto_resource(service_name: str):
+    try:
+        # Track timing for the operation
+        start_time = time.time()
+
+        operation_name = 'Query' if operation_type == 'query' else 'Scan'
+
+        logger.info(f"Starting {operation_name.lower()} for {table_name}...")
+
+        table = dynamodb.Table(table_name)
+
+        # Get the appropriate operation method
+        operation = getattr(table, operation_type)
+
+        # Start fetching items
+        fetch_start = time.time()
+        items = []
+        # Shallow copy to avoid mutating the original during pagination
+        local_operation_kwargs = operation_kwargs.copy()
+        response = operation(**local_operation_kwargs)
+        items.extend(response.get('Items', []))
+
+        # Track pagination
+        page_count = 1
+
+        while 'LastEvaluatedKey' in response:
+            local_operation_kwargs['ExclusiveStartKey'] = response['LastEvaluatedKey']
+            response = operation(**local_operation_kwargs)
+            items.extend(response.get('Items', []))
+            page_count += 1
+
+        fetch_time = time.time() - fetch_start
+        logger.info(f"[{time.time() - start_time:.2f}s] {table_name}: Fetched {len(items)} items across {page_count} pages in {fetch_time:.2f}s")
+
+        # Delete items in batches
+        if items:
+            delete_start = time.time()
+            deleted_count = batch_delete_items(table, items, key_config['keys'])
+            delete_time = time.time() - delete_start
+            logger.info(f"[{time.time() - start_time:.2f}s] {table_name}: Deletion completed in {delete_time:.2f}s")
+
+            total_time = time.time() - start_time
+            logger.info(f"[{time.time() - start_time:.2f}s] {table_name}: Total operation completed in {total_time:.2f}s")
+
+            return deleted_count
+        else:
+            logger.info(f"[{time.time() - start_time:.2f}s] {table_name}: No items to delete")
+            return 0
+
+    except Exception as e:
+        if 'ResourceNotFoundException' in str(e):
+            logger.debug(f"  - {table_name}: Table does not exist")
+            return 0
+        else:
+            operation_name = 'Query' if operation_type == 'query' else 'Scan'
+            logger.error(f"  ✗ {table_name}: {operation_name} failed - {e}")
+            return 0
+
+
+def query_and_delete(dynamodb, table_name, key_config):
+    """
+    Query a DynamoDB table using a partition key and delete matching items.
+
+    Args:
+        dynamodb: DynamoDB resource object
+        table_name: Name of the table to query
+        key_config: Dict with 'keys' (list of key names) and 'partition_value'
+
+    Returns:
+        Number of items deleted
+    """
+    operation_kwargs = {
+        'KeyConditionExpression': Key(key_config['keys'][0]).eq(key_config['partition_value'])
+    }
+    return _fetch_and_delete_items(dynamodb, table_name, key_config, 'query', operation_kwargs)
+
+
+def scan_and_delete_with_prefix(dynamodb, table_name, key_config):
```

Comment:
See previous message - we should never be in a position where this is necessary, thanks to GSI / secondary keys